d3a_interface/sim_results/kpi.py

```diff
@@ -145,11 +145,74 @@ def update_area_kpi(self, area_dict, core_stats):
+        self.consumer_ess_set = set()
+        self.house_list = list()
+        self.fit_revenue = 0.
+        self.utility_bill = 0.
+        self.base_case_revenue = 0.
+        self.d3a_revenue = 0.
+        self.saving_absolute = 0.
+        self.saving_percentage = 0.
+
+    def calculate_saving_kpi(self, area_dict, core_stats, gf_alp):
+        """
+        Args:
+            area_dict: contain nested area info
+            core_stats: contain area's raw/key statistics
+            gf_alp: grid_fee_along_the_path - cumulative grid fee from root to target area
+
+        Returns:
+        """
+        self.populate_area_type(area_dict)
+
+        # fir_excl_gf_alp: feed-in tariff rate excluding grid fee along path
+        fir_excl_gf_alp = self.get_feed_in_tariff_rate_excluding_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        # mmr_incl_gf_alp: market maker rate include grid fee along path
+        mmr_incl_gf_alp = self.get_market_maker_rate_including_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        for trade in core_stats.get(area_dict["uuid"], {}).get("trades", []):
+            if trade["seller_origin_id"] in self.producer_ess_set:
+                self.fit_revenue += fir_excl_gf_alp * trade["energy"]
+                self.d3a_revenue += trade["price"]
+            if trade["buyer_origin_id"] in self.consumer_ess_set and \
+                    trade["seller_origin_id"] not in self.producer_ess_set:
+                self.utility_bill += mmr_incl_gf_alp * trade["energy"]
+                self.d3a_revenue -= trade["price"]
+        self.base_case_revenue = self.fit_revenue - self.utility_bill
+        self.saving_absolute = self.d3a_revenue - self.base_case_revenue
+        self.saving_percentage = ((self.saving_absolute / self.base_case_revenue) * 100
+                                  if self.base_case_revenue else 0.)
+
+    def populate_area_type(self, area_dict):
+        for child in area_dict["children"]:
+            if is_producer_node_type(child):
+                self.producer_ess_set.add(child["uuid"])
+            elif is_load_node_type(child):
+                self.consumer_ess_set.add(child["uuid"])
+            elif is_prosumer_node_type(child):
+                self.producer_ess_set.add(child["uuid"])
+                self.consumer_ess_set.add(child["uuid"])
+
+    @staticmethod
+    def get_feed_in_tariff_rate_excluding_path_grid_fees(area_core_stat, path_grid_fee):
+        return area_core_stat.get("feed_in_tariff", 0.) - path_grid_fee
+
+    @staticmethod
+    def get_market_maker_rate_including_path_grid_fees(area_core_stat, path_grid_fee):
+        return area_core_stat.get("market_maker_rate", 0.) + path_grid_fee
+
+
 class KPI(ResultsBaseClass):
     def __init__(self):
         self.performance_indices = dict()
         self.performance_indices_redis = dict()
         self.state = {}
+        self.saving_state = {}
```

Comment:
Why is this a dict? If I'm not wrong, you're only filling it with one single element:
```python
self.saving_state[area_dict['uuid']] = SavingsKPI()
```
So why don't we instantiate it as `None` here, and do this later?
```python
self.saving_state = SavingsKPI()
```

Reply:
Moreover, the naming of the variable is still inconsistent: it should probably be `savings_state` (with an `s`)

Reply:
> Why is this a dict?

I need to calculate savings kpi for individual areas and all areas calculations are independent of each other.